src/runtime/buffer.h

```diff
@@ -10,8 +10,16 @@ struct buffer {
     void *contents;
 };
 
+/* don't force klibs import symbols for assert */
+#ifdef KLIB
+#define buffer_assert(x)
+#else
+#define buffer_assert(x) assert(x)
+#endif
+
 static inline void *buffer_ref(buffer b, bytes offset)
 {
+    buffer_assert(b->start + offset <= b->length);
```

Comment:
This should be `<` rather than `<=`

Reply:
The problem here is that there is at least foreach_character macro which uses the comma operator in the test to use buffer_ref to set the next value before the test is performed. Let me see if I can rewrite the macro to avoid the buffer_ref if the test will fail.

Reply:
I fixed the foreach_character, but I found another pattern used in a quite a few places is 'buffer_ref(b, buffer_length(b))' to get an end pointer for an iterator. I'm not sure what the best course of action is. Should we make a new method for getting an end pointer, or keep it as is where it could miss off-by-one errors at the end?

Reply:
Sounds good